kube-runtime/src/utils/event_flatten.rs

```diff
@@ -0,0 +1,122 @@
+use crate::watcher::{Error, Event};
+use core::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use futures::{Stream, TryStream};
+use pin_project::pin_project;
+
+// grab from private part of tokio
+macro_rules! ready {
+    ($e:expr $(,)?) => {
+        match $e {
+            std::task::Poll::Ready(t) => t,
+            std::task::Poll::Pending => return std::task::Poll::Pending,
+        }
+    };
+}
+
+#[pin_project]
+/// Stream returned by the [`watch_applies`](super::WatchStreamExt::watch_applies) and [`watch_touches`](super::WatchStreamExt::watch_touches) method.
+#[must_use = "streams do nothing unless polled"]
+pub struct EventFlatten<St, K> {
+    #[pin]
+    stream: St,
+    delete: bool,
+    state: Option<Result<Event<K>, Error>>,
```

Comment:
`std::vec::IntoIter` would allow us to preserve the order from the `Restarted` event and avoid threading already-flat events through the loop

Reply:
For this, we could rewrite the struct as

```rust
pub struct EventFlatten<St, K> {
    #[pin]
    stream: St,
    emit_deleted: bool,
    restarted_queue: std::vec::IntoIter<K>,
}
```

and `poll_next` as

```rust
fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
    let mut me = self.project();
    Poll::Ready(loop {
        if let Some(item) = me.restarted_queue.next() {
            break Some(Ok(item));
        }
        break match ready!(me.stream.as_mut().poll_next(cx)) {
            Some(Ok(Event::Applied(obj))) => Some(Ok(obj)),
            Some(Ok(Event::Deleted(obj))) => {
                if *me.emit_deleted {
                    Some(Ok(obj))
                } else {
                    continue;
                }
            }
            Some(Ok(Event::Restarted(objs))) => {
                *me.restarted_queue = objs.into_iter();
                continue;
            }
            Some(Err(err)) => Some(Err(err)),
            None => return Poll::Ready(None),
        };
    })
}
```

Reply:
Ah, thanks a lot, that is much cleaner. Have ported that in.